public/app/features/dashboard/dashgrid/PanelPerformanceMonitor.tsx

```diff
@@ -0,0 +1,159 @@
+    // overrides are an array of objects, so stringify it all and log changes
+    // in lack of an index, we can't tell which override changed
+    if (oldFieldConfig.overrides !== fieldConfig.overrides) {
+      logEvent(
+        'field config overrides changed',
+        'overrides',
+        JSON.stringify(fieldConfig.overrides),
+        JSON.stringify(oldFieldConfig.overrides)
+      );
+    }
+
+    const oldDefaults: { [key: string]: any } = oldFieldConfig.defaults;
+
+    // go through field config keys except custom, we treat that below
+    for (const [key, value] of Object.entries(fieldConfig.defaults)) {
+      if (key === 'custom') {
+        continue;
+      }
+
+      const newValue: string = typeof value !== 'string' ? JSON.stringify(value) : value;
+      const oldValue: string = typeof value !== 'string' ? JSON.stringify(oldDefaults[key]) : oldDefaults[key];
+
+      if (oldDefaults[key] === undefined) {
+        logEvent('new default field config', key, newValue);
+      } else if (oldValue !== newValue) {
+        logEvent('default field config changed', key, newValue, oldValue);
+      }
+    }
+
+    if (!fieldConfig.defaults.custom) {
+      return;
+    }
+
+    // go through custom field config keys
+    for (const [key, value] of Object.entries(fieldConfig.defaults.custom)) {
+      const newValue: string = typeof value !== 'string' ? JSON.stringify(value) : value;
+      const oldValue: string =
+        typeof value !== 'string' ? JSON.stringify(oldDefaults.custom[key]) : oldDefaults.custom[key];
+
+      if (oldDefaults.custom[key] === undefined) {
+        logEvent('new custom field config', key, newValue);
+      } else if (oldValue !== newValue) {
+        logEvent('custom field config changed', key, newValue, oldValue);
+      }
+    }
+  };
+
+  useEffect(() => {
+    if (!Object.keys(faro).length) {
+      return;
+    }
+
+    requestAnimationFrame(() => {
+      setTimeout(() => {
+        faro.api.pushMeasurement({
+          type: 'internal_panel_measurements_' + props.panelType,
+          values: {
+            start_loading_time_ms: startLoadTime,
+            load_time_ms: performance.now() - startLoadTime,
+          },
+          // should be fixed by https://github.com/grafana/faro-web-sdk/pull/256/
+          // {
+          //   context: {
+          //     panel_type: props.panelType,
+          //   }
+          // }
+        });
+      }, 0);
+    });
+
+    //we don't want to sub for every panel in dashboard, only for one in edit
+    if (props.isInPanelEdit) {
+      const sub = appEvents.subscribe(DashboardSavedEvent, logChanges);
+
+      return () => {
+        //exiting panel edit mode or changing panel type will auto log changes and unsub
+        logChanges();
+        sub.unsubscribe();
+      };
+    }
+
+    return;
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
```

Comment:
is intent of this useEffect to run similar to `componentDidMount` (i.e. run once for setup)?

Reply:
Yes, the hook is set to run once to measure component starting load time and set the dashboard save sub. We could have it so that it measures paint times on every rerender, or when options change, but that might be too much.

Reply:
> the hook is set to run once to measure component starting load time and set the dashboard save sub

"set the dashboard save sub"  ?

Reply:
I meant subscribe to the `DashboardSavedEvent` event. So that `logChanges` runs only when the dashboard is saved.